cpp/src/sampling/detail/sampling_utils_impl.cuh

```diff
@@ -907,5 +907,39 @@ gather_one_hop_edgelist(
   return std::make_tuple(std::move(majors), std::move(minors), std::move(weights));
 }
 
+template <typename vertex_t, typename edge_t, typename weight_t>
+std::tuple<rmm::device_uvector<vertex_t>,
+           rmm::device_uvector<vertex_t>,
+           rmm::device_uvector<weight_t>,
+           rmm::device_uvector<edge_t>>
+count_and_remove_duplicates(raft::handle_t const& handle,
+                            rmm::device_uvector<vertex_t>&& src,
+                            rmm::device_uvector<vertex_t>&& dst,
+                            rmm::device_uvector<weight_t>&& wgt)
+{
+  rmm::device_uvector<edge_t> count(src.size(), handle.get_stream());
+
+  thrust::fill(handle.get_thrust_policy(), count.begin(), count.end(), edge_t{1});
+
+  auto tuple_iter =
+    thrust::make_zip_iterator(thrust::make_tuple(src.begin(), dst.begin(), wgt.begin()));
+
+  auto end_iter = thrust::reduce_by_key(handle.get_thrust_policy(),
+                                        tuple_iter,
+                                        tuple_iter + src.size(),
+                                        count.begin(),
+                                        tuple_iter,
+                                        count.begin());
+
+  size_t sz = thrust::distance(tuple_iter, end_iter.first);
+
+  src.resize(sz, handle.get_stream());
+  dst.resize(sz, handle.get_stream());
+  wgt.resize(sz, handle.get_stream());
+  count.resize(sz, handle.get_stream());
```

Comment:
Call `shrink_to_fit` to free-up memory.

And even better is we can count the number of unique elements and just allocate the exact amount of memory (so no need to resize).

See
```
    auto num_uniques = thrust::count_if(
      rmm::exec_policy(stream_view),
      thrust::make_counting_iterator(size_t{0}),
      thrust::make_counting_iterator(edgelist_majors.size()),
      detail::is_first_in_run_pair_t<vertex_t>{edgelist_majors.data(), edgelist_minors.data()});

    rmm::device_uvector<vertex_t> tmp_edgelist_majors(num_uniques, stream_view);
    rmm::device_uvector<vertex_t> tmp_edgelist_minors(tmp_edgelist_majors.size(), stream_view);
    rmm::device_uvector<weight_t> tmp_edgelist_weights(tmp_edgelist_majors.size(), stream_view);
    thrust::reduce_by_key(rmm::exec_policy(stream_view),
                          pair_first,
                          pair_first + edgelist_majors.size(),
                          (*edgelist_weights).begin(),
                          thrust::make_zip_iterator(thrust::make_tuple(
                            tmp_edgelist_majors.begin(), tmp_edgelist_minors.begin())),
                          tmp_edgelist_weights.begin());
```
https://github.com/rapidsai/cugraph/blob/branch-22.06/cpp/src/structure/coarsen_graph_impl.cuh#L82

Reply:
Change made in next push.